cpp/src/arrow/array/builder_union.cc

```diff
@@ -0,0 +1,62 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/array/builder_union.h"
+
+#include <utility>
+
+#include "arrow/util/logging.h"
+
+namespace arrow {
+
+UnionBuilder::UnionBuilder(MemoryPool* pool,
+                           const std::vector<std::shared_ptr<ArrayBuilder>>& children)
+    : ArrayBuilder(nullptr, pool), types_builder_(pool), offsets_builder_(pool) {
+  children_ = children;
+}
+
+Status UnionBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
+  std::shared_ptr<Buffer> types;
+  RETURN_NOT_OK(types_builder_.Finish(&types));
+  std::shared_ptr<Buffer> offsets;
+  RETURN_NOT_OK(offsets_builder_.Finish(&offsets));
+
+  std::vector<std::shared_ptr<Field>> fields;
+  std::vector<std::shared_ptr<ArrayData>> child_data(children_.size());
+  std::vector<uint8_t> type_ids;
+  for (size_t i = 0; i < children_.size(); ++i) {
+    if (children_[i]) {
+      std::shared_ptr<ArrayData> data;
+      RETURN_NOT_OK(children_[i]->FinishInternal(&data));
+      child_data[i] = data;
+      fields.push_back(::arrow::field("", children_[i]->type()));
+      type_ids.push_back(static_cast<uint8_t>(i));
+    } else {
+      fields.push_back(::arrow::field("", ::arrow::null()));
```

Comment:
Yeah you are right! It's also better from an efficiency point of view since we don't need all these extra child buffers then (if the types aren't there) and potentially the type metadata can be smaller. One problem then is to identify the children types properly (i.e. which child index corresponds to an INT), but I think we might be able to use https://github.com/apache/arrow/blob/master/format/Schema.fbs#L92 to describe the mapping between child types and the PythonType tags.

I'll give that a shot.